WooCommerce/Classes/Analytics/WaitingTimeTracker.swift

```diff
@@ -0,0 +1,106 @@
+import Foundation
+import Combine
+
+class WaitingTimeTracker {
+    private let waitingTimeout: TimeInterval
+
+    @Published private(set) var currentState: State = .idle
+
+    private var waitingTimer: Timer? = nil
+    private var subscriptions: Set<AnyCancellable> = []
+
+    /// Initialize the WaitingTimeTracker with a specific timeout, if none is provided it will set 30 seconds as the default
+    ///
+    init(waitingTimeout: TimeInterval = 30) {
+        self.waitingTimeout = waitingTimeout
+        configureCurrentState()
+    }
+
+    /// Configure the state changes and react to the specified one
+    /// - `.done`: Will send the elapsed waiting time to the Tracks
+    /// - `.waiting`: Will trigger the timeout Timer and wait for the `.done` state to sink
+    /// - `.idle`: Does nothing and breaks the state changing cycle
+    ///
+    private func configureCurrentState() {
+        $currentState
+            .removeDuplicates()
+            .sink { [weak self] state in
+                switch state {
+                case .done(let waitingEndedTime):
+                    self?.sendWaitingTimeToTracks(
+                            waitingEndedTime: waitingEndedTime
+                    )
+                    self?.resetTrackerState()
+                case .waiting:
+                    self?.startWaitingTimer()
+                case .idle:
+                    break
+                }
+            }
+            .store(in: &subscriptions)
+    }
+
+    /// Set the Tracker state to `.waiting`, triggering the waiting cycle
+    ///
+    /// - parameter analyticsStat: The stat to be send to Tracks when the waiting time ends
+    ///
+    func onWaitingStarted(analyticsStat: WooAnalyticsStat) {
+        currentState = .waiting(NSDate().timeIntervalSince1970, analyticsStat)
+    }
+
+    /// Set the Tracker state to `.done`, ending the waiting cycle. Only accepts it if the current state is `.waiting`
+    /// otherwise, it will ignore the call
+    ///
+    func onWaitingEnded() {
+        guard case .waiting = currentState else {
+            return
+        }
+
+        currentState = .done(NSDate().timeIntervalSince1970)
```

Comment:
```suggestion
        currentState = .done(Date().timeIntervalSince1970)
```

We should prefer native `Swift` types, rather than legacy Objc ones!